charmcraft/application/commands/remote.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""Build a charm remotely on Launchpad."""
+import argparse
+import os
+import pathlib
+import time
+from collections.abc import Collection
+from typing import Any, cast
+
+from craft_application.commands import ExtensibleCommand
+from craft_application.launchpad.models import Build, BuildState
+from craft_application.remote.utils import get_build_id
+from craft_cli import emit
+from overrides import override  # pyright: ignore[reportUnknownVariableType]
+
+from charmcraft import utils
+
+OVERVIEW = """\
+Command remote-build sends the current project to be built
+remotely.  After the build is complete, packages for each
+architecture are retrieved and will be available in the
+local filesystem.
+
+Interrupted remote builds can be resumed using the --recover
+option, followed by the build number informed when the remote
+build was originally dispatched. The current state of the
+remote build for each architecture can be checked using the
+--status option.
+
+To set a timeout on the remote-build command, use the option
+``--launchpad-timeout=<seconds>``. The timeout is local, so
+the build on launchpad will continue even if the local instance
+is interrupted or times out.
+"""
+
+_CONFIRMATION_PROMPT = (
+    "All data sent to remote builders will be publicly available. "
+    "Are you sure you want to continue?"
+)
+
+
+class RemoteBuild(ExtensibleCommand):
```

Comment:
I made this an `ExtensibleCommand` because I think we could probably upstream it to `craft-application` and add app-dependent functionality.